extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,580 @@
+
+  /**
+   * Registers this instance as the host API for the current sentinel.
+   * If the global safeframe listener has not yet been created, it creates
+   * that as well.
+   */
+  registerSafeframeHost() {
+    safeframeHosts[this.sentinel_] = safeframeHosts[this.sentinel_] || this;
+    if (!safeframeListenerCreated) {
+      safeframeListenerCreated = true;
+      this.win_.addEventListener('message', safeframeListener, false);
+    }
+  }
+
+  /**
+   * Sends initial connection message to the safeframe to finish initialization.
+   * Also initializes the sending of geometry update messages to the frame.
+   */
+  connectMessagingChannel(channel) {
+    // Set the iframe here, because when class is first created the iframe
+    // element does not yet exist on this.baseInstance_. The first time
+    // we receive a message we know that it now exists.
+    this.iframe_ = this.baseInstance_.iframe;
+    this.channel = channel;
+    dev().assert(this.iframe_.contentWindow,
+        'Frame contentWindow unavailable.');
+    this.setupGeom_();
+    this.sendMessage_(dict({
+      'message': 'connect',
+      'c': this.channel,
+    }), '');
+  }
+
+  /**
+   * Creates IntersectionObserver instance for this SafeframeAPI instance.
+   * We utilize the existing IntersectionObserver class, by passing in this
+   * class for IO to use instead of SubscriptionApi for sending its update
+   * messages. The method 'send' below is triggered by IO every time that
+   * an update occurs.
+   * @private
+   */
+  setupGeom_() {
+    this.IntersectionObserver_ = new IntersectionObserver(
+        this.baseInstance_, this.baseInstance_.element, false, this, 1000);
+    // This will send a geometry update message, and then start sending
+    // them whenever geometry changes as detected by intersection
+    // observer. Make sure we always send this initial update message.
+    this.IntersectionObserver_.startSendingIntersectionChanges();
+  }
+
+  /**
+   * Every time that the IntersectionObserver needs to send an update, this
+   * method is triggered. Includes page scroll or other visibility change
+   * events.
+   * Handles sending the geometry update message to the
+   * safeframe container, which allows $sf.ext.geom() to work.
+   */
+  send(unusedTrash, changes) {
+    this.sendMessage_({
+      newGeometry: this.formatGeom_(changes['changes'][0]),
+      uid: this.uid,
+    }, SERVICE.GEOMETRY_UPDATE);
+  }
+
+  /**
+   * The IntersectionObserver class is monitoring the amp-ad element,
+   * not the Safeframe. The Safeframe is not necessarily the exact
+   * same size as the amp-ad element, so calculate the width and height
+   * correction between the amp-ad element and the safeframe iframe, and then
+   * modify the change entry to be correct.
+   * @param {!Object} changes
+   * @return {!Object} Corrected intersection change entry.
+   */
+  correctChanges(changes) {
+    const iframeRect = this.iframe_ ? this.iframe_.getBoundingClientRect() :
```

Comment:
This is going to require review given you're calling getBoundingClientRect.  It may be fine but we'll want AMP core to review.  Can we calculate and know this as iframe attachment time given we center the creative?  Are we concerned about pubs who put CSS that overrides our centering assuming we're not using !important?